Add UO SRML data reader

lboeman opened the pull request:
pvlib python pull request guidelines
====================================

Thank you for your contribution to pvlib python! You may delete all of these instructions except for the list below.

You may submit a pull request with your code at any stage of completion.

The following items must be addressed before the code can be merged. Please don't hesitate to ask for help if you're unsure of how to accomplish any of the items below:

 - [x] Closes #589 
 - [x] I am familiar with the [contributing guidelines](http://pvlib-python.readthedocs.io/en/latest/contributing.html).
 - [x] Fully tested. Added and/or modified tests to ensure correct behavior for all reasonable inputs. Tests (usually) must pass on the TravisCI and Appveyor testing services.
 - [x] Updates entries to `docs/sphinx/source/api.rst` for API changes.
 - [x] Adds description and name entries in the appropriate `docs/sphinx/source/whatsnew` file for all changes.
 - [x] Code quality and style is sufficient. Passes LGTM and SticklerCI checks.
 - [x] New code is fully documented. Includes sphinx/numpydoc compliant docstrings and comments in the code where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.

Brief description of the problem and proposed solution (if not already fully described in the issue linked to above):

wholmgren commented on pvlib/iotools/srml.py:
```diff
@@ -0,0 +1,148 @@
+    Spectral data (7xxx) uses all four digits to indicate the
+    variable.
+    """
+    var_map = {
```
this is ok but I am wondering if it would be better to define it at the top of the module. Then it's easier for people to inspect it if they are interested.

lboeman replied:
I agree, I must have misunderstood something the other day, I'll replace this at the top of the module.

wholmgren commented on pvlib/iotools/srml.py:
```diff
@@ -0,0 +1,148 @@
+        except KeyError:
+            return col
+    try:
+        return var_map[col[:3]] + '_' + col[3:]
```
recommend more explicit code like

```
variable_name = var_map[col[:3]]
variable_number = col[3:]
return variable_name + '_' + variable_number
```

wholmgren commented on pvlib/iotools/srml.py:
```diff
@@ -0,0 +1,148 @@
+
+
+def request_srml_data(station, year, month, filetype='PO'):
+    """Read a month of SRML data from solardat into a Dataframe.
```
This needs the same references as the basic read function above.

wholmgren commented on pvlib/iotools/srml.py:
```diff
@@ -0,0 +1,148 @@
+    return df
+
+
+def request_srml_data(station, year, month, filetype='PO'):
```
maybe `read_srml_from_solardat` and import it in `__init__.py`?

lboeman replied:
Should this module include these kind of util functions? I wasn't sure about including this here, or a function to stitch together due to their function signatures being pretty far off from read_<data_type> functions.

wholmgren replied:
I think the module should include any generally useful functions for IO related to the UOSRML. That would include downloading directly from their website. 

I suppose this function might more accurately be called `read_srml_month_from_solardat`

I'm less sure about a function to download and stitch together multiple months because you said that the columns sometimes change. If we made it, that's the function that would be called `read_srml_from_solardat`

lboeman commented on pvlib/iotools/srml.py:
```diff
@@ -0,0 +1,148 @@
+    # associated var so we create a dict mapping them to var_flag for renaming
+    columns = data.columns
+    flag_label_map = {flag: columns[columns.get_loc(flag) - 1] + '_flag'
+                      for flag in columns[1::2]}
```
I should be more descriptive with this, pandas appends a `.#` value to the duplicates, so we end up with a series like [0.1, 0.2, 0.3, 0.4].
The resulting dict is something like:
```{ '0.1' : 'ghi_0_flag',
      '0.2' : 'temp_air_0_flag',
    }```

lboeman commented on pvlib/iotools/srml.py:
```diff
@@ -0,0 +1,148 @@
+    data = data.rename(columns=flag_label_map)
+    # For data flagged bad or missing, replace the value with np.NaN
+    for col in columns[::2]:
+        data[col] = data[col].where(~(data[col + '_flag'] == 99), np.NaN)
```
I found that -99.9 and -999 both existed as fill values for non-existant data. However, the flag values for 'bad or missing' is 99 regardless of the data value so I'm using that here to set all missing data values to NaN.

wholmgren replied:
My mistake. I see now that you're testing the flag column, not the data column.

lboeman commented:
Updated in response to some of your comments:
- I've added references to the University of Oregon Solar Radiation Monitoring Laboratory to read_srml and read_srml_month_from_solardat. If there is a better format for those attributions, I can update them properly.
- I had originally included a full month of data for testing, but I created a smaller file with only a day of data called SRML-day-EUPO1801.txt. The file only includes three variables to make it easier to scan.
- Added comments for clarity around flag parsing code.Updated in response to some of your comments:
- I've added references to the University of Oregon Solar Radiation Monitoring Laboratory to read_srml and read_srml_month_from_solardat. If there is a better format for those attributions, I can update them properly.
- I had originally included a full month of data for testing, but I created a smaller file with only a day of data called SRML-day-EUPO1801.txt. The file only includes three variables to make it easier to scan.
- Added comments for clarity around flag parsing code.

Would it be useful if this could handle data other than 1-min resolution?

lboeman commented on pvlib/iotools/srml.py:
```diff
@@ -0,0 +1,183 @@
+    """
+    df_time = df[df.columns[1]] - 1
+    df_doy = df[df.columns[0]]
+    fifty_nines = df_time % 100 == 99
```
These variable names might be more confusing than I had intended. 59 is the desired result here, after reducing the times by 1 all of the hours become HH99(e.g. 2400->2399), so they need to be adjusted. 
This was to get around exceptions trying to parse a datetime from day 366 hour 2400 of a leap year. It is admittedly quite confusing. I will document the rationale better here.

cwhanse replied:
Aha, makes sense now.  Because of -1, the times run from hh00, hh01, ..., hh58, hh99.  Minor nit - putting `df_doy = ` ahead of  `df_time = ` would help me see the connection.

wholmgren commented on pvlib/iotools/srml.py:
```diff
@@ -0,0 +1,187 @@
+    Returns
+    -------
+    df: pd.Dataframe
+        The Dataframe with a datetime index applied.
```
DatetimeIndex localized to 'Etc/GMT+8'.

wholmgren commented on pvlib/iotools/srml.py:
```diff
@@ -0,0 +1,187 @@
+    filetype: string
+        SRML file type to gather. 'RO' and 'PO' are the
+        only minute resolution files.
+    Returns
```
blank line above returns

wholmgren commented on pvlib/iotools/srml.py:
```diff
@@ -0,0 +1,187 @@
+    ----------
+    [1] University of Oregon Solar Radiation Measurement Laboratory
+        `http://solardat.uoregon.edu/ <http://solardat.uoregon.edu/>`_
+
```
but no blank line here

wholmgren commented on pvlib/iotools/srml.py:
```diff
@@ -0,0 +1,187 @@
+        month=month)
+    url = "http://solardat.uoregon.edu/download/Archive/"
+    data = read_srml(url + file_name)
+    return data
```
empty line at end of file

lboeman replied:
flake8 complains about this. with
`srml.py:185:1: W391 blank line at end of file`
Should I add it anyway? From what i can tell, pep8 just wants a newline at the end of the file and not an extra blank line.

wholmgren commented on pvlib/test/test_srml.py:
```diff
@@ -0,0 +1,75 @@
+    assert data.index[0] == start
+    assert data.index[-1] == end
+    assert (data.index[59::60].minute == 59).all()
+    assert year not in data.columns
```
can this assert fail? it's type sensitive...

```python
In [3]: df = pd.DataFrame(columns=['2016'])

In [4]: '2016' in df.columns
Out[4]: True

In [5]: 2016 in df.columns
Out[5]: False

In [6]: df = pd.DataFrame(columns=[2016])

In [7]: '2016' in df.columns
Out[7]: False

In [8]: 2016 in df.columns
Out[8]: True
```

wholmgren commented:
great, thanks @lboeman
